File: src/bmsx_cpp/machine/runtime/engine_irq.cpp
#include "machine/runtime/engine_irq.h"

#include "machine/bus/io.h"
#include "machine/runtime/runtime.h"
#include "machine/runtime/runtime_fault.h"

#include <stdexcept>

namespace bmsx {
namespace {

constexpr uint32_t ENGINE_IRQ_MASK = IRQ_REINIT | IRQ_NEWGAME;

} // namespace

void raiseEngineIrq(Runtime& runtime, uint32_t mask) {
	if (mask == 0u) {
		throw new Error("engine IRQ mask must be non-zero.");
	}
	const uint32_t unsupported = mask & ~ENGINE_IRQ_MASK;
	if (unsupported != 0u) {
		throw new Error("unsupported engine IRQ mask " + std::to_string(unsupported) + ".");
	}
	runtime.machine().irqController().raise(mask);
}

} // namespace bmsx

File: src/bmsx/machine/runtime/engine_irq.ts
import { IRQ_NEWGAME, IRQ_REINIT } from '../bus/io';
import type { Runtime } from './runtime';

const ENGINE_IRQ_MASK = (IRQ_REINIT | IRQ_NEWGAME) >>> 0;

export function raiseEngineIrq(runtime: Runtime, mask: number): void {
	const normalized = mask >>> 0;
	if (normalized === 0) {
		throw new Error('engine IRQ mask must be non-zero.');
	}
	const unsupported = normalized & ~ENGINE_IRQ_MASK;
	if (unsupported !== 0) {
		throw new Error(`unsupported engine IRQ mask 0x${unsupported.toString(16)}.`);
	}
	runtime.machine.irqController.raise(normalized);
}
